changelog/copy-auto-ref.dd

```diff
@@ -0,0 +1,35 @@
+`std.algorithm.mutation.copy` Has Been Modified For `put`-style Output Ranges
+
+With this release, when the target output range in a call to
+$(REF copy, std, algorithm, mutation) is any `put` defining output range (and it
+doesn't have $(REF_ALTTEXT assignable elements, hasAssignableElements, std, range, primitives)),
+the target is taken as `auto ref`. This new overload is also `void` returning
+vs the other overloads which return the unfilled portion of the given range
+or array.
+
+Before this release, $(REF copy, std, algorithm, mutation) took all
+$(REF_ALTTEXT output ranges, isOutputRange, std, range, primitives) by value.
+This leads to some issues when output ranges aren't designed around a reference
+to heap memory, e.g. the hash ranges in $(MREF std,digest):
+
+-------
+// v2.078 and earlier
+import std.algorithm.mutation : copy;
+import std.digest.digest : toHexString;
+import std.digest.md : MD5;
+import std.range : put;
+
+auto s = "Hello!\n";
+auto h1 = makeDigest!MD5;
+auto h2 = makeDigest!MD5;
+
+put(h1, s); // takes h1 by ref
+copy(s, h2); // takes h2 by value
+
+assert(h1.finish().toHexString == "E134CED312B3511D88943D57CCD70C83");
+assert(h2.finish().toHexString == "D41D8CD98F00B204E9800998ECF8427E");
+// this fails in 2.079 and later
+assert(h1.finish().toHexString != h2.finish().toHexString)
```

Comment:
Are you allowed to call `finish` more than once? I'm not familiar with the API of digests. Also the assert is trivially obvious (the two strings are different per the above asserts).

I think in the suggestion I was saying, you could just put a note after the code saying the two digests will be the same in 2.079 and later.

Alternatively, you could leave out the 3rd assert and put as a comment: "In 2.079 and later, both digests will have the same value".